### PR TITLE
[RLlib] Preserve PBT-perturbed hyperparameters across checkpoint restore

### LINKED ISSUE #62173: [RLlib|Tune] PBT-perturbed hyperparameters are silently ignored after checkpoint restore in PPO
### What happened + What you expected to happen

When using Population-Based Training (PBT) with PPO, perturbed hyperparameters are only
reflected at the `Algorithm` level. After an exploit step -- where a weaker trial clones a
stronger trial's checkpoint and applies perturbations -- the trial is restored from that
checkpoint, which **overwrites the perturbed values** with the cloned trial's originals. As a
result, `clip_param`, `lambda`, and lr` (and possibly others) are never actually applied during training.

## Root cause

PBT correctly updates `algorithm.config` (via `trial.set_config()`), but the checkpoint
restore path undoes this in two places:

**1. `policy.py` -- `Policy.set_state()` (lines 1062-1063):**

```python
if policy_spec.config:
    self.config = policy_spec.config  # overwrites perturbed values
```

This unconditionally replaces `policy.config` with the serialized config from the checkpoint.
Since PPO reads `clip_param` and `lambda` directly from `policy.config` (in the loss function
and GAE postprocessing respectively), those perturbations are silently discarded.

**2. `torch_policy_v2.py` -- `TorchPolicyV2.set_state()` (lines 1048-1052):**

```python
optim_state_dict = {"param_groups": s["param_groups"]}  # includes lr
o.load_state_dict(optim_state_dict)  # restores lr from checkpoint
```

Restoring `param_groups` also restores `lr` in the optimizer, overriding the PBT-perturbed
learning rate. There is no subsequent correction unless an `lr_schedule` is configured.

### Versions / Dependencies

- Ray 2.10.0
- Python 3.11.14
- ...

### Reproduction script

```python
"""
PBT + RLlib bug: Policy.set_state() overwrites perturbed hyperparameters.

After PBT exploit (checkpoint copy + HP perturbation), Policy.set_state()
restores policy.config from the checkpoint, undoing perturbations for lr,
clip_param, and lambda_. Only algorithm-level params like num_sgd_iter survive.

Root cause (ray/rllib/policy/policy.py:1030-1031):
    if policy_spec.config:
        self.config = policy_spec.config

Additionally, optimizer.load_state_dict() restores lr from checkpoint.

Tested on Ray 2.10. Same code exists on latest master.
"""

import os
os.environ["PYTHONWARNINGS"] = "ignore::DeprecationWarning"

import ray
from ray import tune
from ray.rllib.algorithms.ppo import PPOConfig
from ray.rllib.algorithms.callbacks import DefaultCallbacks
from ray.tune.schedulers import PopulationBasedTraining


class HPInspector(DefaultCallbacks):
    def on_train_result(self, *, algorithm, result, **kwargs):
        policy = algorithm.get_policy()
        result["algo_lr"] = algorithm.config["lr"]
        result["pol_lr"] = policy._optimizers[0].param_groups[0]["lr"]
        result["algo_clip"] = algorithm.config["clip_param"]
        result["pol_clip"] = policy.config["clip_param"]
        result["algo_lambda"] = algorithm.config["lambda_"]
        result["pol_lambda"] = policy.config["lambda"]


ray.init(num_cpus=4, ignore_reinit_error=True, logging_level="ERROR")

pbt = PopulationBasedTraining(
    time_attr="training_iteration",
    metric="episode_reward_mean",
    mode="max",
    perturbation_interval=1,
    quantile_fraction=0.5,
    hyperparam_mutations={
        "lr": tune.uniform(1e-5, 1e-2),
        "clip_param": tune.uniform(0.1, 0.5),
        "lambda": tune.uniform(0.5, 1.0),
    },
    synch=True,
)

config = (
    PPOConfig()
    .environment("CartPole-v1")
    .training(lr=5e-4, clip_param=0.3, lambda_=0.95, train_batch_size=500)
    .rollouts(num_rollout_workers=0)
    .callbacks(HPInspector)
)

results = tune.run(
    "PPO",
    config=config.to_dict(),
    scheduler=pbt,
    num_samples=2,
    stop={"training_iteration": 6},
    verbose=0,
)

# Print comparison table
fmt = "{:<12} {:<5} {:>10} {:>10} {:>6} {:>10} {:>10} {:>6} {:>10} {:>10} {:>6}"
print(fmt.format(
    "Trial", "Iter",
    "algo_lr", "pol_lr", "lr?",
    "algo_clip", "pol_clip", "clip?",
    "algo_lam", "pol_lam", "lam?",
))
print("-" * 105)

mismatches = 0
for trial in results.trials:
    df = None
    for key, value in results.trial_dataframes.items():
        if trial.trial_id in key or key in trial.trial_id:
            df = value
            break
    if df is None:
        continue

    for _, row in df.iterrows():
        if "algo_lr" not in row:
            continue

        lr_ok = abs(row["algo_lr"] - row["pol_lr"]) < 1e-9
        clip_ok = abs(row["algo_clip"] - row["pol_clip"]) < 1e-6
        lam_ok = abs(row["algo_lambda"] - row["pol_lambda"]) < 1e-6

        print(fmt.format(
            trial.trial_id, int(row["training_iteration"]),
            f"{row['algo_lr']:.6f}", f"{row['pol_lr']:.6f}", "ok" if lr_ok else "FAIL",
            f"{row['algo_clip']:.4f}", f"{row['pol_clip']:.4f}", "ok" if clip_ok else "FAIL",
            f"{row['algo_lambda']:.4f}", f"{row['pol_lambda']:.4f}", "ok" if lam_ok else "FAIL",
        ))

        if not (lr_ok and clip_ok and lam_ok):
            mismatches += 1

print("-" * 105)
if mismatches:
    print(f"BUG: {mismatches} rows where algorithm.config != policy.config after PBT exploit.")
    print("Policy.set_state() overwrites perturbed config with checkpoint values.")
else:
    print("No mismatches (PBT may not have exploited). Try running again.")

ray.shutdown()
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by joaquinhuigomez**
## Why are these changes needed?

PBT correctly perturbs hyperparameters on `algorithm.config` via `trial.set_config()`, but checkpoint restore in `Policy.set_state()` unconditionally replaces `policy.config` with the cloned trial's serialized config, silently discarding `clip_param`, `lambda`, and `lr`. Additionally, `TorchPolicyV2.set_state()` restores optimizer `param_groups` from the checkpoint, overwriting the perturbed learning rate.

This PR merges the checkpoint config under the current policy config (so PBT-perturbed values take precedence) and re-applies the current `lr` to optimizer param groups after state restoration.

Fixes #62173

## Related issue number

#62173

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
Thanks for the PR @joaquinhuigomez
Gemini has noted a number of issues, I would guess that the problem is actually upstream of what you've changed. 
Is it clear why these configs are being missed when setting the state? Have the configs been lost or are just ignored?

**Comment by joaquinhuigomez** [comment]
Fair point on the upstream investigation needed. Closing this for now.
